src/Storages/MergeTree/MergeTreeReadPoolParallelReplicas.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <Core/NamesAndTypes.h>
+#include <Storages/MergeTree/MergeTreeBaseSelectProcessor.h>
+#include <Storages/MergeTree/MergeTreeBlockReadUtils.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+#include <Storages/MergeTree/MergeTreeReadPool.h>
+#include <Storages/MergeTree/RangesInDataPart.h>
+#include <Storages/SelectQueryInfo.h>
+#include "Storages/MergeTree/RequestResponse.h"
+#include "Storages/StorageSnapshot.h"
+
+#include <condition_variable>
+#include <mutex>
+#include <memory>
+
+namespace DB
+{
+
+
+/// Weird class with copy-paste from original read pool
+class MergeTreeReadPoolParallelReplicas :  private boost::noncopyable
```

Comment:
Would be good to see what's the difference with an ordinary pool.
Now, looks like a lot of code was copy-pasted.

Reply:
This pool doesn't split all the tasks. Everything it does - just collect the info about all the parts for further usage and act like a proxy for requests to coordinator. It could be done inside ordinary pool, but I guess it will make the code completely unreadable. For isolation purposes I decided to create a separate class.